src/account/PluginManagerInternals.sol

```diff
@@ -86,11 +86,11 @@ abstract contract PluginManagerInternals is IPluginManager {
     {
         SelectorData storage _selectorData = getAccountStorage().selectorData[selector];
 
-        if (_selectorData.validation.notEmpty()) {
+        // Fail on duplicate definitions - otherwise dependencies could shadow non-depdency
+        // validation functions, leading to partial uninstalls.
+        if (!_selectorData.validations.add(toSetValue(validationFunction))) {
```

Comment:
Validation routing is done off-chain so definitely worth exploring moving execution selector routing off-chain too!

Re: session keys, i think the decision is between plugin + 1-key-per-plugin (or multiple keys that are managed by a dapp) v.s. 1 plugin + n keys per plugin?
1. I do prefer the security model of 1 key per plugin since we can use the account's permissions framework directly. That's preferable for 2 reasons - we don't need to trust the plugin to manage individual key permissions, and we dedupe permissions checking (in the other model, account allocates a global limit to the plugin, then the plugin allocates per-key limits). This plugin could be as lightweight as a signature validator if the native permissions framework is sufficiently granular/flexible
3. But in the 1-key-per-plugin model, we'll potentially have to deploy a new contract to add a new key (probably a proxy from a ProxyFactory, which is the safe/zodiac modules model). Unless the deployed plugin is ultra lightweight it should be more costly than the other approach

Reply:
Yeah we definitely need some changes to support that workflow, but I wanted to keep each change scoped down to just what is needed per PR.

Also, I think having 1 key per plugin is too inefficient - we can treat the function id as a per-account, per-key item to perform that switching, like in the validator experiments.

"Key" is also the wrong term here, it's leaking the underlying validation logic into a higher layer of abstraction. They're really independent "validation functions" or "validators".

Reply:
> Also, I think having 1 key per plugin is too inefficient - we can treat the function id as a per-account, per-key item to perform that switching, like in the validator experiments.

Hmm I might be missing something, but here's an example I was thinking about - in the case of adding a new session key to the existing MA session key plugin with `n` permissions, the minimal state change required will be: 
1. `n` x nonzero to nonzero `SSTORE`s via increment the ERC20 + native token spend limits from the account to the plugin
2. `n` x zero to nonzero `SSTORE`s via incrementing the ERC20 + native token spend limits of the plugin to the session key

Whereas deploying a new validator would be just `n` x zero to nonzero `SSTORE`s via incrementing the ERC20 + native token spend limits of the account to the new validator

Reply:
I know in the old v0.7 model, we basically needed two layers of permissions - account <> plugin, and in the case of a key-based validation plugin, then permissions for each key.

If we switch to composable validation (aka, allow validation plugins to be installed multiple times), then we go back to just needing 1 set of permissions per "validation", which is a combination of both the address of the validator plugin and the ID of the key within it.

I think the model of doing proxies on plugins per account is wrong - the cost of any contract creation is already too. high (32k), not to mention the added forwarding cost of being a proxy.